Add App Check token to FirebaseServerApp

DellaBitta opened the pull request:
### Discussion

Currently a customer enforces App Check for a product in the Firebase console then there's no way to successfully use that product in SSR environments. This is due to the requirement for products SDKs to internally invoke `getToken` on the a instance of the App Check SDK that the customer's app has initialized. However, the App Check SDK can only be initialized in browser environments.

This PR aims to fix this. FirebaseServerApp will now accept an optional App Check token at initialization. The product SDKs will look for this token, and if it's present, the SDKs will use this value in lieu of calling `getToken` on App Check.

This change affects the following SDKs:
 * Auth
 * Cloud Functions
 * Data Connect
 * Firestore
 * Real Time Database
 * Vertex AI

### Testing

Added tests to FirebaseServerApp's handling of the token. The product SDKs currently don't have viable App Check tests in our test suite, so I manually tested all of the SDKs by enabling App Check enforcement in the Firebase Console, watching them fail due to permissions, and then provided the App Check token to Firebase Server app and watched them succeed.

### API Changes

This conforms to the previoulsy approved [Existing FirebaseServerApp API proposal](http://goto.google.com/fb-web-FirebaseServerApp) (internal) for `FirebaseServerApp` which includes App Check token support.

DellaBitta commented on packages/data-connect/src/core/AppCheckTokenProvider.ts:
```diff
@@ -29,10 +30,14 @@ import { Provider } from '@firebase/component';
   private appCheck?: FirebaseAppCheckInternal;
+  private serverAppAppCheckToken?: string;
   constructor(
-    private appName_: string,
```
This value was never used.

DellaBitta commented on packages/data-connect/src/core/AppCheckTokenProvider.ts:
```diff
@@ -29,10 +30,14 @@ import { Provider } from '@firebase/component';
+    if (_isFirebaseServerApp(app) && app.settings.appCheckToken) {
+      this.serverAppAppCheckToken = app.settings.appCheckToken;
+    }
     this.appCheck = appCheckProvider?.getImmediate({ optional: true });
```
During initialization of the `AppCheckTokenProvider` that Data Connect implements, check to see if the provided instance of `FirebaseApp` is one of `FirebaseServerApp`, and if it contains an App Check token, then store it locally so that it can be returned by this provide when its `getToken` method is invoked (below).

DellaBitta commented on packages/data-connect/src/core/AppCheckTokenProvider.ts:
```diff
@@ -29,10 +30,14 @@ import { Provider } from '@firebase/component';
   ) {
+    if (_isFirebaseServerApp(app) && app.settings.appCheckToken) {
+      this.serverAppAppCheckToken = app.settings.appCheckToken;
+    }
```
During initialization of the `AppCheckTokenProvider` that Data Connect implements, check to see if the provided instance of `FirebaseApp` is one of `FirebaseServerApp`, and if it contains an App Check token, then store it locally so that it can be returned by this provide when its `getToken` method is invoked (below).

DellaBitta commented on packages/database/src/core/AppCheckTokenProvider.ts:
```diff
@@ -30,17 +31,26 @@ import { warn } from './util/util';
+    this.appName = app.name;
+    if (_isFirebaseServerApp(app) && app.settings.appCheckToken) {
+      this.serverAppAppCheckToken = app.settings.appCheckToken;
+    }
```
During initialization of the `AppCheckTokenProvider` that RTDB implements, check to see if the provided instance of `FirebaseApp` is one of `FirebaseServerApp`, and if it contains an App Check token, then store it locally so that it can be returned by this provide when its `getToken` method is invoked (below).

DellaBitta commented on packages/firestore/src/api/credentials.ts:
```diff
@@ -495,10 +496,16 @@ export class FirebaseAppCheckTokenProvider
+    if (_isFirebaseServerApp(app) && app.settings.appCheckToken) {
+      this.serverAppAppCheckToken = app.settings.appCheckToken;
+    }
+  }
```
Similiar to the other SDKs (above) during the initialization of the `AppCheckTokenProvider` that irestore implements, check to see if the provided instance of `FirebaseApp` is one of `FirebaseServerApp`, and if it contains an App Check token, then store it locally so it can be returned by the `getToken` method (below).

tom-andersen replied:
Is there any concern of this token expiring?

DellaBitta replied:
I attempted to answer this in the main thread of the PR.

DellaBitta commented on packages/functions/src/context.ts:
```diff
@@ -47,11 +48,16 @@ export class ContextProvider {
   ) {
+    if (_isFirebaseServerApp(app) && app.settings.appCheckToken) {
+      this.serverAppAppCheckToken = app.settings.appCheckToken;
+    }
```
Similiar to the other SDKs (above) during the initialization of the `AppCheckTokenProvider` that functions implements, check to see if the provided instance of `FirebaseApp` is one of `FirebaseServerApp`, and if it contains an App Check token, then store it locally so it can be returned by the `getToken` method (below).

DellaBitta commented on packages/storage/src/service.ts:
```diff
@@ -262,6 +266,9 @@ export class FirebaseStorageImpl implements FirebaseStorage {
   async _getAppCheckToken(): Promise<string | null> {
+    if (_isFirebaseServerApp(this.app) && this.app.settings.appCheckToken) {
+      return this.app.settings.appCheckToken;
+    }
```
Similiar to the other SDKs (above) check to see if the provided instance of `FirebaseApp` is one of `FirebaseServerApp`, and if it contains an App Check token, then return it instead of invoking the App Check SDK.

DellaBitta commented on packages/auth/test/integration/flows/firebaseserverapp.test.ts:
```diff
@@ -166,37 +166,6 @@ describe('Integration test: Auth FirebaseServerApp tests', () => {
-
-    await deleteApp(serverApp);
-  });
-
```
Invalid tokens are now handled by a new flow, which we test in `app/src/FirebaseServerApp.test.ts`, so I'm removing this test here.

MarkDuckworth commented on packages/database/src/core/AppCheckTokenProvider.ts:
```diff
@@ -30,17 +31,26 @@ import { warn } from './util/util';
     }
   }
 
   getToken(forceRefresh?: boolean): Promise<AppCheckTokenResult> {
```
if `forceRefresh && this.serverAppAppCheckToken` should this throw or otherwise indicate that an unsupported call was made?

DellaBitta replied:
Just checking: is the only time that `forceRefresh` is set to `true` for the App Check token is when the service returns an `UNAUTHENTICATED` error?

hsubox76 replied:
It looks like, in RTDB, it will be set to true on any failed request (status !== 'ok') to the auth or app check endpoints.

DellaBitta replied:
I suppose that throwing here would then be valid, as it would reduce the burden on the service on subsequent requests, if the app attempts to continue to use the defunct App Check Token. WDYT @hsubox76 ?

maneesht replied:
LGTM from my end

MarkDuckworth reviewed (approved):
Looks okay. One comment. Also, we we're wondering if/how this solution deals with the expiration or short lifespan of app check tokens?

DellaBitta commented:
> Looks okay. One comment. Also, we we're wondering if/how this solution deals with the expiration or short lifespan of app check tokens?

App Check tokens last a minimum of 30 minutes. It's assumed that they will be created via `getToken` on the client-side and then passed up to the SSR phase via cookies or service workers, probably at first at client sign in.

We'll post a best practice that apps attempt to use that App Check token on the server side, and either check the validity of the token themselves, or call `initializeServerApp` with a `try` / `catch` block to check for token expiration. If the server side encounters an expired token then the app should route to a client-side page that regenerates a new token, and then route back to the page that uses SSR.

hsubox76 commented on packages/functions/src/context.ts:
```diff
@@ -47,11 +48,16 @@ export class ContextProvider {
   ) {
+    if (_isFirebaseServerApp(app) && app.settings.appCheckToken) {
+      this.serverAppAppCheckToken = app.settings.appCheckToken;
+    }
```
Can't attach comment to line 85 but can we put a `?` there so it's `appCheckProvider?.get()` like RTDB and DataConnect etc seem to have done and I just noticed in this PR, for max efficiency in not bothering to even call `get()` in the case of no provider.

DellaBitta replied:
Done.

hsubox76 commented on packages/vertexai/src/models/generative-model.ts:
```diff
@@ -82,7 +83,17 @@ export class GenerativeModel {
-      if ((vertexAI as VertexAIService).appCheck) {
+
+      if (
+        vertexAI.app &&
```
Can there be a vertexAI instance that doesn't have an `app` property? How does that happen? Actually come to think of it, what do you think about putting `a` ? here https://github.com/firebase/firebase-js-sdk/blob/main/packages/app/src/internal.ts#L173 so that it can handle being passed undefined/null, so it would cover if any other caller happens to pass it that.

DellaBitta replied:
To answer your first question, I'm not sure. I don't think it's possible but the rest of the code (above) checks for it.

As for the latter, I'll add the ability for the function to take null / undefined!

DellaBitta replied:
Done.

hsubox76 commented on packages/vertexai/src/models/generative-model.ts:
```diff
@@ -82,7 +83,17 @@ export class GenerativeModel {
+        this._apiSettings.getAppCheckToken = () => {
+          return Promise.resolve({ token });
+        };
+      } else if ((vertexAI as VertexAIService).appCheck) {
```
I don't think this PR is the time to do it, but we should probably abstract all this logic (maybe also the model name processing) out of `GenerativeModel` to some kind of shared function that ImagenModel can also use, will give @dlarocque a heads up.

hsubox76 reviewed (approved):
LG, just nits!

NhienLam reviewed (approved):
Auth: LGTM

DellaBitta commented on packages/data-connect/src/core/AppCheckTokenProvider.ts:
```diff
@@ -42,7 +47,11 @@ export class AppCheckTokenProvider {
   }
 
-  getToken(forceRefresh?: boolean): Promise<AppCheckTokenResult> {
+  getToken(): Promise<AppCheckTokenResult> {
```
forceRefresh was never used, so I've removed it for now.

egilmorez commented on packages/app/src/firebaseServerApp.ts:
```diff
@@ -67,6 +93,16 @@ export class FirebaseServerAppImpl
       ...serverConfig
     };
 
+    // Ensure that the current time is within the authIdtoken window of validity.
```
authIdtoken and appChecktoken look like literals that should be backticked.

DellaBitta replied:
Fixed.

egilmorez commented on packages/app/src/public-types.ts:
```diff
@@ -196,6 +196,12 @@ export interface FirebaseServerAppSettings
   authIdToken?: string;
 
+  /**
+   * An optional App Check token. If provided, the Firebase SDKs that use App Check will utilizze
```
Typo "utilize."

Personally I like in lieu, but I'll bet good money our style guide advises something more like "instead of" or "in place of."

DellaBitta replied:
Fixed. Updated the comments for the new parts again, so it's worth a full review from the begining if you don't mind.

egilmorez reviewed (commented):
Couple of things to look at, thanks!

jamesdaniels commented on packages/firestore/src/api/credentials.ts:
```diff
@@ -562,6 +569,9 @@ export class FirebaseAppCheckTokenProvider
   }
 
   getToken(): Promise<Token | null> {
+    if (this.serverAppAppCheckToken !== null) {
```
probably just want truthy here to handle `undefined`

jamesdaniels replied:
it should be find with just null, as that's what it's defaulted to but it seems a foot gun

DellaBitta replied:
Done!

jamesdaniels commented on packages/app/src/firebaseServerApp.test.ts:
```diff
@@ -155,4 +170,142 @@ describe('FirebaseServerApp', () => {
+    expect(encounteredError).to.be.false;
+  });
+
+  it('throws when authIdToken has expired', () => {
```
per our approved design, expired/invalid tokens should console.error and treat as undefined, rather than throw. expired tokens, especially for AppCheck are expected in normal SSR operation as their TTL is very short. it should be up to the developer to handle them

DellaBitta replied:
Updated, PTAL!

DellaBitta commented:
> Couple of things to look at, thanks!

Thanks @egilmorez. Could you give it a review again?  I fixed the problems you found and updated some other comments, so it's worth a full review again.

egilmorez commented on .changeset/kind-pets-sin.md:
```diff
@@ -0,0 +1,15 @@
+'@firebase/auth': patch
+---
+
+`FirebaseServerApp` may now be initalized with an App Check token in leu of invoking the App Check
```
Do these strings get pulled into release notes or other public docs? 

If so, I'd go with "can now be initialized" and "instead of invoking."

DellaBitta replied:
They don't directly. Later, the release engineer crafts the release notes using these as ... inspiration. Updated!

egilmorez reviewed (approved):
LG with one final possible nit, thanks!
